profiler/src/trace_reader.rs

```diff
@@ -77,26 +122,73 @@ pub fn collect_samples_from_trace(trace: &CallTrace) -> Vec<Sample> {
     samples
 }
 
-fn collect_samples(
+pub struct ResourcesKeys {
+    pub builtins: HashSet<String>,
+    pub syscalls: HashSet<DeprecatedSyscallSelector>,
+}
+
+impl ResourcesKeys {
+    pub fn value_types(&self, context: &mut ProfilerContext) -> Vec<ValueType> {
+        let mut value_types = vec![];
+
+        for builtin in &self.builtins {
+            value_types.push(ValueType {
+                r#type: context.string_id(builtin).into(),
+                unit: context.string_id(&String::from("count")).into(),
+            });
+        }
+        for syscall in &self.syscalls {
+            value_types.push(ValueType {
+                r#type: context.string_id(&format!("{syscall:?}")).into(),
+                unit: context.string_id(&String::from("count")).into(),
+            });
+        }
+
+        value_types
+    }
+}
+
+pub fn collect_resources_keys(samples: &[Sample]) -> ResourcesKeys {
+    let mut syscalls = HashSet::new();
+    let mut builtins = HashSet::new();
+    for sample in samples {
+        builtins.extend(
+            sample
+                .flat_resources
+                .vm_resources
+                .builtin_instance_counter
+                .keys()
+                .cloned(),
+        );
+        syscalls.extend(sample.flat_resources.syscall_counter.keys())
+    }
+    ResourcesKeys { syscalls, builtins }
+}
```

Comment:
This doesn't really require iteration rn (would be enough to get the keys from the top call), but since blockifier plans to change resources to not include inner calls we will prob change the semantics of the expected input file as well - then this iteration will be needed.
 
TL;DR: not needed, but more flexibile, doesn't do any harm